kratos/elements/levelset_convection_element_simplex_algebraic_stabilization.cpp

```diff
@@ -0,0 +1,220 @@
+        const auto& rGradVar = p_conv_diff_settings->GetGradientVariable();
+
+        //getting data for the given geometry
+        BoundedMatrix<double, TNumNodes, TDim > DN_DX;
+        array_1d<double, TNumNodes > N;
+        double Volume;
+
+        const auto& r_geom = this->GetGeometry();
+        GeometryUtils::CalculateGeometryData(r_geom, DN_DX, N, Volume);
+
+        //here we get all the variables we will need
+        array_1d<double,TNumNodes> phi, phi_old;
+        array_1d< array_1d<double,3 >, TNumNodes> v, vold;
+
+        array_1d<double,3 > X_mean_tmp = ZeroVector(3);
+        array_1d<double,3 > grad_phi_mean_tmp = ZeroVector(3);
+        array_1d< array_1d<double,3 >, TNumNodes> X_node;
+        double phi_mean_old = 0.0;
+        double phi_mean = 0.0;
+
+        for (unsigned int i = 0; i < TNumNodes; ++i)
+        {
+            const auto& r_node = r_geom[i];
+
+            phi[i] = r_node.FastGetSolutionStepValue(rUnknownVar);
+            phi_old[i] = r_node.FastGetSolutionStepValue(rUnknownVar,1);
+
+            v[i] = r_node.FastGetSolutionStepValue(rConvVar);
+            vold[i] = r_node.FastGetSolutionStepValue(rConvVar,1);
+
+            X_mean_tmp += r_node.Coordinates();
+            X_node[i] = r_node.Coordinates();
+
+            grad_phi_mean_tmp += r_node.GetValue(rGradVar);
+
+            phi_mean_old += r_node.FastGetSolutionStepValue(rUnknownVar,1);
+            phi_mean += r_node.FastGetSolutionStepValue(rUnknownVar);
+        }
+
+        phi_mean /= static_cast<double>(TNumNodes);
+        phi_mean_old /= static_cast<double>(TNumNodes);
+
+        array_1d<double,TDim> X_mean, grad_phi_mean;
+        for(unsigned int k = 0; k < TDim; k++)
+        {
+            grad_phi_mean[k] = grad_phi_mean_tmp[k]/static_cast<double>(TNumNodes);
+            X_mean[k] = X_mean_tmp[k]/static_cast<double>(TNumNodes);
+        }
+
+        array_1d<double,TDim> grad_phi_diff = prod(trans(DN_DX), phi_old) - grad_phi_mean;
+
+        BoundedMatrix<double,TNumNodes, TNumNodes> K_matrix = ZeroMatrix(TNumNodes, TNumNodes); // convection
+        BoundedMatrix<double,TNumNodes, TNumNodes> S_matrix = ZeroMatrix(TNumNodes, TNumNodes); // LHS stabilization
+        Vector S_vector = ZeroVector(TNumNodes); // RHS stabilization
+        BoundedMatrix<double,TNumNodes, TNumNodes> Mc_matrix = ZeroMatrix(TNumNodes, TNumNodes); // consistent mass matrix
+        BoundedMatrix<double,TNumNodes, TNumNodes> Ml_matrix = IdentityMatrix(TNumNodes, TNumNodes); // lumped mass matrix
+
+        BoundedMatrix<double,TNumNodes, TNumNodes> M_matrix;
+        BoundedMatrix<double,TNumNodes, TNumNodes> L_matrix;
+
+        BoundedMatrix<double,TNumNodes, TNumNodes> Ncontainer;
+        this->GetShapeFunctionsOnGauss(Ncontainer);
+
+        array_1d<double, TDim > vel_gauss;
+        array_1d<double, TDim > X_gauss;
+        array_1d<double, TNumNodes > v_dot_grad_N;
+
+        const double limiter = this->GetValue(LIMITER_COEFFICIENT);
```

Comment:
I guess this value is constant for all the elements. If this is the case it must be stored in the `ProcessInfo` for the sake of memory consumption. If not dismiss this comment.

Reply:
It varies element by element.